<!--
  Launch configuration for an app manager in multimaster mode. This exposes all
  the bells and whistles available for configuration. See concert_client.launch
  for an example which tunes it for a particular usage.
 -->
<launch>
  <!-- ******************************* Common Args ***************************** -->
  <arg name="robot_name"                    default="Cybernetic Pirate"/>

  <!-- ******************************** Gateway ******************************** -->
  <arg name="gateway_firewall"          default="true" doc="default is not to let anything come into the robot" />
  <arg name="gateway_uuids"             default="true" doc="all gateways represent their robots with unique uuid's postfixed to the name" />
  <arg name="gateway_watch_loop_period" default="10" doc="the polling period for multimaster advertising/flipping" />
  <arg name="hub_whitelist"             default="[]" doc="list of hub names/regex patterns" />
  <arg name="hub_uri"                   default=""/>
  <arg name="disable_zeroconf"          default="false"/>

  <include file="$(find rocon_app_manager)/launch/includes/_gateway.xml">
    <arg name="gateway_name" value="$(arg robot_name)" />
    <arg name="gateway_firewall" value="$(arg gateway_firewall)" />
    <arg name="disable_uuids" value="true" unless="$(arg gateway_uuids)" />
    <arg name="gateway_watch_loop_period" value="$(arg gateway_watch_loop_period)" />
    <arg name="hub_whitelist" value="$(arg hub_whitelist)" />
    <arg name="hub_uri" value="$(arg hub_uri)"/>
    <arg name="disable_zeroconf"  value="$(arg disable_zeroconf)"/>
  </include>

  <!-- ****************************** Capabilities ****************************** -->
  <arg name="capabilities"                       default="false" doc="enable/disable a capability server" />
  <arg name="capabilities_parameters"            default="$(find rocon_app_manager)/param/capabilities.yaml" doc="detailed parameter configuration for the providers" />
  <arg name="capabilities_blacklist"             default="[]" doc="blacklist specific capabilities" />
  <arg name="capabilities_nodelet_manager_name"  default="capability_server_nodelet_manager" />
  <arg name="capabilities_package_whitelist"     default="[std_capabilities]" doc="get capabilities from these packages only" />
  <arg name="capabilities_server_name"           default="capability_server"/>

  <group if="$(arg capabilities)">
    <include file="$(find rocon_app_manager)/launch/includes/_capabilities.xml">
      <arg name="blacklist"            value="$(arg capabilities_blacklist)" />
      <arg name="nodelet_manager_name" value="$(arg capabilities_nodelet_manager_name)" />
      <arg name="name"                 value="$(arg capabilities_server_name)" />
      <arg name="package_whitelist"    value="$(arg capabilities_package_whitelist)" />
      <arg name="parameters"           value="$(arg capabilities_parameters)" />
    </include>
  </group>

  <!-- ****************************** Rapp Manager ****************************** -->
  <arg name="auto_rapp_installation"        default="false" doc="http://wiki.ros.org/rocon_app_manager/Tutorials/indigo/Automatic Rapp Installation" />
  <arg name="auto_start_rapp"               default="" doc="autostart a rapp, e.g. rocon_apps/chirp" />
  <arg name="local_remote_controllers_only" default="false" doc="allow remote control for local machine tests only" />
  <arg name="rapp_package_whitelist"        default="[rocon_apps]" doc="comma separated list of package names" />
  <arg name="rapp_package_blacklist"        default="[]"/>
  <arg name="rapp_preferred_configuration_file" default="$(find rocon_app_manager)/param/preferred_default.yaml"/>
  <arg name="robot_type"                    default="pc"/>
  <arg name="robot_icon"                    default="rocon_icons/cybernetic_pirate.png"/>
  <arg name="screen"                        default="false" doc="verbose output from running apps" />
  <arg name="simulation"                    default="false" doc="if simulated robot" />

  <include file="$(find rocon_app_manager)/launch/includes/_app_manager.xml">
    <arg name="robot_name" value="$(arg robot_name)" />
    <arg name="robot_type" value="$(arg robot_type)" />
    <arg name="robot_icon" value="$(arg robot_icon)" />
    <arg name="use_gateway_uuids" value="$(arg gateway_uuids)" />
    <arg name="rapp_package_whitelist" value="$(arg rapp_package_whitelist)" />
    <arg name="rapp_package_blacklist" value="$(arg rapp_package_blacklist)" />
    <arg name="rapp_preferred_configuration_file" value="$(arg rapp_preferred_configuration_file)" />
    <arg name="auto_start_rapp" value="$(arg auto_start_rapp)" />
    <arg name="local_remote_controllers_only" value="$(arg local_remote_controllers_only)" />
    <arg name="screen" value="$(arg screen)" />
    <arg name="simulation" value="$(arg simulation)"/>
    <arg name="capability_server_name" value="$(arg capabilities_server_name)" if="$(arg capabilities)"/>
    <arg name="auto_rapp_installation" value="$(arg auto_rapp_installation)" />
  </include>

  <!-- ***************************** Rocon Master Info ************************** -->
  <arg name="robot_description"      default="To err is human, to 'arr is pirate."/>

  <include file="$(find rocon_app_manager)/launch/includes/_rocon_master_info.xml">
    <arg name="name" value="$(arg robot_name)" />
    <arg name="description" value="$(arg robot_description)" />
    <arg name="icon" value="$(arg robot_icon)" />
  </include>

  <!-- ***************************** Interactions ******************************* -->
  <arg name="interactions"          default="false"/>
  <arg name="interactions_list"     default="[]"/>

  <group if="$(arg interactions)">
    <include file="$(find rocon_app_manager)/launch/includes/_interactions.xml">
      <arg name="interactions_list" value="$(arg interactions_list)"/>
    </include>
  </group>

  <!-- ******************************* Zeroconf ******************************** -->
  <!-- We might like to advertise our ros master sometimes here, like for standalone -->
  <group unless="$(arg disable_zeroconf)">
    <node ns="zeroconf" pkg="zeroconf_avahi" type="zeroconf" name="zeroconf"/>
  </group>

</launch>
